### PR TITLE
Feature/similarity data

### PR DESCRIPTION
**Pull request by pdan93**
This implements the functionality needed to create similar pretraining/training data based on a given dataset.

Features:

1. construct_embedding_data() - that can construct new embeddings based on different model class given to it.
2. get_similar_data() - the function that will use a given dataset and a set of embeddings to create the similar data.
3. helper functions & classes that make different components easy to change.

### CODE REVIEW

**Comment by pdan93** [comment]
So I tried to make this as easy to use in a more general way. In the interest of that I think hosting the set of embeddings already created would help as the generator can just download one of those at a time and get_similar_data will use them to create the similar data. 
So this should help in the future to easily augment training data or create pretraining data by just having a dataset with text&label columns and calling the default version of the get_similar_data function.

File: mantis/similar.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    def __init__(self):
+        self.model = hub.load("https://tfhub.dev/google/universal-sentence-encoder/4")
+
+    def get_embeddings(self, data):
+        return self.model(data).numpy()
+
+
+def MantisGutenbergMiniLM():
+    """
+    Generator that downloads the minilm embedding files one at a time
```
Comment (nsorros) [inline mantis/similar.py:40]: This seems a bit too specific, can you explain what is the intention? Have you already transformed the data into vectors and retrieving them? if so, with which model and how does it apply to other projects?

File: requirements.txt [hunk 2]
```diff
@@ -1,3 +1,9 @@
 pytest
 typer
 spacy
+transformers
```
Comment (nsorros) [inline requirements.txt:4]: I understand this predates your addition but it would be great to move to more reproducible builds. I think we aim to use this as a library not as an application and as such we possibly need to maintain only a setup.py. In `setup.py` it is more acceptable to have unpinned versions

File: mantis/cli.py [hunk 3]
```diff
@@ -1,12 +1,90 @@
...
+    similar_data = get_similar_data(
+        data_items, 
+        labels=data_labels, 
+        threshold=threshold,
+        embeddings_generator=embeddings_generator,
+        model_encoder_class=model_encoder_class
+    )
+
+    similar_data_df = pd.DataFrame(similar_data)
+    similar_data_df.to_csv(out_csv)
```
Comment (nsorros) [inline mantis/cli.py:87]: if the only thing we use pandas for is to read and write csv it would possibly better and maybe more efficient to use the `csv` module

**Comment by nsorros** [comment]
> This implements the functionality needed to create similar pretraining/training data based on a given dataset.
> 
> Features:
> 
> 1. construct_embedding_data() - that can construct new embeddings based on different model class given to it.
> 2. get_similar_data() - the function that will use a given dataset and a set of embeddings to create the similar data.
> 3. helper functions & classes that make different components easy to change.

It would be useful to provide an example in your description of these are meant to be used

File: mantis/cli.py [hunk 4]
```diff
@@ -1,12 +1,90 @@
...
+    Parameters:
+    txt_tile (string): Filename of the textfile. File should have on each separate line the desired data
+    out_dir (string): Directory name where the embedding json files will be created
+    character_length_threshold (int): lower limit for nr of characters each item should have
+    model_encoder_class_name (string): If you want to use a custom model, just implement the the MantisEncoder and make it available in the classpath, then you can specify the name here
+    """
+
+    model_encoder_class = SentenceTransformerEncoder
+    if model_encoder_class_name != "SentenceTransformerEncoder":
+        model_encoder_class = getattr(sys.modules[__name__], model_encoder_class_name)
```
Comment (nsorros) [inline mantis/cli.py:34]: this seems a bit dodgy, what are you trying to do and what does the value end up being?

File: mantis/cli.py [hunk 5]
```diff
@@ -1,12 +1,90 @@
...
+        txt_file, 
+        out_dir=out_dir, 
+        model_encoder_class=model_encoder_class,
+        character_length_threshold=character_length_threshold
+    )
+    
+
+@app.command()
+def get_similar_training_data(
+    data_csv: str = "clrp_train.csv",
```
Comment (nsorros) [inline mantis/cli.py:46]: this should not have a default value

File: mantis/similar.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    Returns:
+    Will create the embedding json files for the given text data. Each embedding file has 2 lists:
+    paragraphs - the original text data
+    embeddings - the embedding vector for the respective text item
+    """
+
+    model_encoder = model_encoder_class()
+    vectors_per_chunk = int(vectors_per_file/10)
+
+    embedding_data = {
```
Comment (nsorros) [inline mantis/similar.py:108]: note that this is not memory efficient, you can iterate and write on the file in a JSONL format without storing in memory, something like
```
with open("data.jsonl") as f:
    for paragraph, embedding in yield_paragraph_embedding(...):
        f.write(json.dumps({"paragraph": paragraph, "embedding": embedding}))
        f.write("\n")
```

File: mantis/similar.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    Returns:
+    list of dict: The similar data in a list of dicts, if labels are present the dict will also include the 'label' key
+    """
+
+    model_encoder = model_encoder_class()
+    data_embeddings = model_encoder.get_embeddings(data)
+
+    generator = embeddings_generator()
+    similar_data = []
+    while True:
```
Comment (nsorros) [inline mantis/similar.py:185]: I think you can simply iterate the generator
```
for embedding_data in generator:
    ...
```
no need for while

File: mantis/similar.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+                    print("Processed set {} in {} seconds".format(doc_nr, time_taken))
+    if verbose:
+        print("Finished constructing embedding data")
+
+
+def get_similar_data(
+        data, 
+        labels=None, 
+        threshold=0.7, 
+        embeddings_generator=MantisGutenbergMiniLM,
```
Comment (nsorros) [inline mantis/similar.py:163]: the `generator` suffix is superfluous and possibly confusing, just embeddings is enough

Comment (nsorros) [reply mantis/similar.py:163]: this should be the second positional argument and it should not accept a default possibly.

**Review by nsorros — CHANGES_REQUESTED** [review]
This is a good start but there are a couple of points that need addressing and I am still thinking about the API overall. Some points for the time being are
* remove the data, remove the egg
* more memory efficient creation for embeddings
* storing the embeddings in some cache that is checked

To be fair though this assumes the API remains as is which I am not 100% sure so let me review again tomorrow.

**Comment by ivyleavedtoadflax** [comment]
> What is the `clrp_train.csv`? Does it need to be part of the PR? Also you probably need to ignore the egg that is added

Yep i think that needs to go somewhere else

File: mantis/cli.py (hunk 3 above)
Comment (ivyleavedtoadflax) [reply mantis/cli.py:87]: https://twitter.com/a_campello_jr/status/1418173014273843207

File: mantis/similar.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            "mantisnlp-kaggle-data",
+            "minilm_embeddings_{}.json".format(doc_i),
+            'minilm_embeddings.json'
+        )
+        with open("minilm_embeddings.json", 'r') as f:
+            embedding_data = json.load(f)
+            embedding_data['embeddings'] = np.array(embedding_data['embeddings'])
+            yield embedding_data
+
+def get_gutenberg_text():
```
[the remark this answers is not in the document]
Comment (ivyleavedtoadflax) [reply mantis/similar.py:61]: Agree. There's a pattern for doing this in https://github.com/wellcometrust/WellcomeML. We also have a pattern in mantis for doing this using dvc, that i could also dig out.

**Comment by nsorros** [comment]
Ok so I had a think about it and here is where I am.

The `get_similar_training_data` function should receive two positional paths, one to the labeled data and one to the bigger unlabelled data along with some keyword arguments for labels, threshold and encoder as you have. It should iterate the unlabelled data, calculate the embedding and cache it somewhere (possibly another argument) so that if embeddings are calculated there there is no need to recalculate them.

The `make_embedding_data` function can stay the same as it serves the purpose of caching the embeddings.

So the main difference is the generator which at the time retrieves the embeddings from s3. It is fine to retrieve from s3 if the embeddings are there and store somewhere locally in a cache. If they are not there, it should calculate them locally. There should also be a push function to push them to s3. The other difference is as I mentioned in the `get_similar_data` where I believe that the unlabelled big dataset should be provided which in turn would be passed to the generator and the generator will deal with whether this is cached and if not it can check whether it is in s3 or has to calculate it.

To make matter more readable note that you can pass in strings as well as paths as arguments for the datasets and maintain a dict where you map some of those strings to paths. For example 
```
DATASETS = {
    'guttenberg': '~/.cache/mantis/guttenberg/'
}

def get_embedding_data(data.csv, dataset):
    if dataset in DATASETS:
        dataset = DATASETs[dataset]
```

File: mantis/cli.py (hunk 4 above)
Comment (pdan93) [reply mantis/cli.py:34]: I'm trying to somehow give the ability to use another class that implements the MantisEncoder interface. I admit this solution isn't the best and it was actually intended to be up for discussion. 
Right now this is the way I found, of basically the classname be specified in the param, and this should load that class. I would love to go for a cleaner way but couldn't really find any. any ideas?

File: mantis/similar.py (hunk 6 above)
Comment (pdan93) [reply mantis/similar.py:108]: I actually didn't know that was possible, to iterate over a jsonl file. cool, will do that

File: mantis/similar.py (hunk 7 above)
Comment (pdan93) [reply mantis/similar.py:185]: yes, will change. Honestly, the while originally had sense there cause there was a different setup, but yes, now it's not needed.

**Comment by pdan93** [comment]
> Ok so I had a think about it and here is where I am.
> 
> The `get_similar_training_data` function should receive two positional paths, one to the labeled data and one to the bigger unlabelled data along with some keyword arguments for labels, threshold and encoder as you have. It should iterate the unlabelled data, calculate the embedding and cache it somewhere (possibly another argument) so that if embeddings are calculated there there is no need to recalculate them.
> 
> The `make_embedding_data` function can stay the same as it serves the purpose of caching the embeddings.
> 
> So the main difference is the generator which at the time retrieves the embeddings from s3. It is fine to retrieve from s3 if the embeddings are there and store somewhere locally in a cache. If they are not there, it should calculate them locally. There should also be a push function to push them to s3. The other difference is as I mentioned in the `get_similar_data` where I believe that the unlabelled big dataset should be provided which in turn would be passed to the generator and the generator will deal with whether this is cached and if not it can check whether it is in s3 or has to calculate it.
> 
> To make matter more readable note that you can pass in strings as well as paths as arguments for the datasets and maintain a dict where you map some of those strings to paths. For example
> 
> ```
> DATASETS = {
>     'guttenberg': '~/.cache/mantis/guttenberg/'
> }
> 
> def get_embedding_data(data.csv, dataset):
>     if dataset in DATASETS:
>         dataset = DATASETs[dataset]
> ```

I get the general idea of where we want to go with it. But I have some questions:
1. originally I did those 2 functions cause to me the construction of the vectors seemed as important as the construction of the similar data. Do you want to move from that to putting the main focus only on the similar data?
2. that generator thing was supposed to be like an example of how I need the data yielded, I get that it's a bit custom, what if I make an interface class that will be used for that? (and my generator for that specific embedding can then implement that)
3. Initially I did these default values thinking that the kaggle version will be the default. But I guess this is intended to be a helper library, and so no default is better then a custom one huh? I'll move everythign that's kaggle task related in a different file to make it clear it's not part of the main structure here.
4. Regarding files. As you see now everything is in this similar.py file, and that is not my intention, but I had some problems making the library work when I try to create a separate folder in there. It doesn't see the path and just get weird errors. (so for example if I would create 'gutenberg' folder in mantis, and in it have files for the MantisGutenbergMiniLM thing. and then try in similar.py to do `from gutenberg.generator import MantisGutenbergMiniLM` I get the error that gutenberg is not a module. And yes of course I already tried debugging, making sure __init__.py is everywhere, etc..) You know something I don't about this part?

**Comment by nsorros** [comment]
> 4. Regarding files. As you see now everything is in this similar.py file, and that is not my intention, but I had some problems making the library work when I try to create a separate folder in there. It doesn't see the path and just get weird errors. (so for example if I would create 'gutenberg' folder in mantis, and in it have files for the MantisGutenbergMiniLM thing. and then try in similar.py to do `from gutenberg.generator import MantisGutenbergMiniLM` I get the error that gutenberg is not a module. And yes of course I already tried debugging, making sure **init**.py is everywhere, etc..) You know something I don't about this part?

Starting from last, I suggest when you get errors you open an issue or ask in our stack overflow or even slack. This is the best way to resolve them.

 >1. originally I did those 2 functions cause to me the construction of the vectors seemed as important as the construction of the similar data. Do you want to move from that to putting the main focus only on the similar data?

well constructing the vectors is relatively trivial isn't it? we are just using pretrained models.

> 2. that generator thing was supposed to be like an example of how I need the data yielded, I get that it's a bit custom, what if I make an interface class that will be used for that? (and my generator for that specific embedding can then implement that)

If we want to share the Kaggle code we should create a separate repo. The intention here is to create some generalised functions which you can use for Kaggle and other projects.

> 3. Initially I did these default values thinking that the kaggle version will be the default. But I guess this is intended to be a helper library, and so no default is better then a custom one huh? I'll move everythign that's kaggle task related in a different file to make it clear it's not part of the main structure here.

defaults are good see sklearn where it is part of the design to not need to pass any arguments to instantiate a class. but even in sklearn `fit` has to have an `X` and `Y` so in our case we need to decide what is a default and what should be passed. the data that are labeled and the corpus from which we are drawing additional examples I think should be passed. to be honest you can argue that there is a default corpus from which we can draw examples kind of like how bert default is trained in wikipedia but is Gutenberg that default? maybe but I am not sure.
